### ISSUE TITLE
Random function not completely random when using multiple root streams

### ISSUE THREAD

**Issue by JesperJanssen** [mvysny/photocloud-frame-slideshow#36]
V 1.9.6

I have the feeling that certain photo's apear more often than others. 

In my stream I have 3 root streams selected. 1 root stream has 20 sub folders (and a total of 15000 photo's) the other 2 root streams have both only 2 sub folders (with both a total of 2500 photo's). When using the random option for the slideshow I see photo's from the subfolders in the root streams with only 2 sub folders more often as from the subfolders in the root stream with 20 folders. Is this just the way the random function works or is this some kind of bug? 

If it was complete random than 12.5% of the photo displayed should come from the root stream with two sub folders (2500 photos) and another 12.5% from the other root stream with two subfolders (2500 photos) and 75% from the root stream with 20 subfolders (15000 photos).

See below for a stand which I created for my tablet which is running this app (screensaver function) constantly.
![20170901_184047](https://user-images.githubusercontent.com/17165733/29979706-dd157876-8f46-11e7-992b-118a3756ed42.jpg)

**Comment by mvysny** [comment]
Hey, I really love that stand, really awesome, good-looking, non-intrusive, good work :-) :+1: 
The dirty secret is that when the slideshow starts, PhotoCloud does not know how many files there are and thus employs a very simple algorithm: 1 photo from the first root, 1 photo from the second root, etc. So initially the probability is 33% when having three roots. Only after the 2500-photo roots are "drained", then the pictures from the 15000 photos will only be shown until they are all shown.
Hard to say whether this algorithm can be improved, please feel free to suggest improvements ;)

**Comment by JesperJanssen** [comment]
So with three root streams (A, B, C) you would see photos in the following order ABCABCABC. However, I do often see multiple photos from 1 root stream after each other. Does the amount of sub folders in each root stream not influence the algorithm?

See below for a picture of the stand when I touch the display (screensavers turns off). It then changes into a dashboard for my domtica system.

**Comment by mvysny** [comment]
Domtica system: awesome idea, I love it! So PhotoCloud is a screensaver, replaced by the "house control panel" when you touch it? Smart ;)

Streams: exactly. If you have three streams and the photos would be shown in sequential order, then A is first polled for a photo, then B, then C, then A again for a next photo, etc, generating the ABCABCABC sequence. After A and B runs out of photos, the sequence is turned into CCCC once C runs out of photos too, then the whole thing is repeated.

However, there is one catch: the photo order is randomized. So not only A, B and C are walked in random order, the order of the photos in the triplet is randomized, too. Thus you can get something like ACB BAC ACB BAC ABC etc. Note that it is possible that two photos from B may be shown consecutively. Yet perhaps this is not wanted: perhaps only the walk order should be randomized, but the photos from the streams should still be shown in the strictly ABCABCABC sequence? What do you think?

**Comment by JesperJanssen** [comment]
At the moment I think it is more random as the ABCABC option I describe above. So I would only change the random option if there is a way that the amount of subfolders in each root stream is taken into account (take into account the total amount of photos is not practical and I think would be worse). So if the app would know that root stream 1 has three subfolders ABC and root stream 2 has three subfolders DEF and root stream 3 has 5 sub folders GHIJK than the random order could be to just display one random photo of each of those sub folders. So the order would be something like this: ABCDEFGHIJK or KJABIHCDGFE or etc. And once one sub folder is completely drained you could skip that folder for the next serie or start with that folder over again.

Don't know if this is better or worse as the random option is now. But it is just an idea you can close this issue of you think it will not be an improvement ;)

**Comment by mvysny** [comment]
Unfortunately because of how the system works, that would cause parallel photo download from the same stream, which would probably slow down the download itself; also I'm not sure how those cloud services like such parallel downloads.

Yet, I have two things in mind:

1. The algorithm can be changed to avoid the CCCCC scenario: all iterators will automatically repeat.
2. I can add a configuration option to show the full name of the image, including the stream it came from. Then you can exactly verify and see the PhotoCloud's behavior, whether it actually follows the ABC permutations, or whether it's showing some stream way more often than the other.

**Comment by mvysny** [comment]
Hmm, I'm not sure whether I understand completely - now that I reread the original bug, there may be a misunderstanding in terminology. There are two things:

1. A Source - the cloud service which provides photos, such as OneCloud
2. A Source root - you can limit PhotoCloud to show only photos from a set of directories in given Source.

If you only have one Source with three Roots, then the slideshow should be completely random - it should do a randomized breadth-first structure iteration which is random.

On the other hand if you have more than one Source, then the above ABC ABC CCC applies.

**Comment by mvysny** [comment]
So I'm guessing that in the beginning of the cycle, the less deeply nested photos are discovered sooner, that's why you could see them sooner and more frequently. However when the entire directory structure is examined, then the photo stream should be balanced (since the dir crawler adds all discovered photos to one list in random order). 

Yet, the balance may be broken if the photo download is slow - since it has higher priority than dir crawling, the dir crawling might be stalled and thus more deeply nested photos are discovered later on...

**Comment by JesperJanssen** [comment]
Indeed, I have one source (One Drive) with three roots. So the slideshow should be completely random and it should not matter if one root contains 2 subfolders and the other 20?

But perhaps it takes some time, as you explained, for the slideshow to be completely random. In the beginning of the slideshow the roots with 2 subfolders will be shown more often since those were faster to load.

What do you mean with dir crawling? Is the app creating an order to display the photo's in an temporary file and once the slideshow is ended this order will be deleted and a new one will be created once it is started again.

**Comment by mvysny** [comment]
> Indeed, I have one source (One Drive) with three roots. So the slideshow should be completely random and it should not matter if one root contains 2 subfolders and the other 20?
>
> But perhaps it takes some time, as you explained, for the slideshow to be completely random. In the beginning of the slideshow the roots with 2 subfolders will be shown more often since those were faster to load.

Exactly as you say. Once the folders are completely loaded and examined, then the random should be completely fair. The longer the pause between photos, the more time PhotoCloud has to load all dirs and the sooner the photo stream will became completely random.

> What do you mean with dir crawling? Is the app creating an order to display the photo's in an temporary file and once the slideshow is ended this order will be deleted and a new one will be created once it is started again.

Dir crawling is just a tech term for examining folders for subfolders and files; to walk over the entire folder/file hierarchy and to find all photo images. Images found are added to the "order" list (which is held in memory) at random order, and then the Slideshow follows this "order" list.

**Comment by JesperJanssen** [comment]
But once all folders are completely loaded will the beginning of the random order be changed? Because the beginning of the display order was created when the random order was not completely random yet and because of that every time when the slideshow restarts the photo less deeply nested photos will be in the front of the random que.

What do you mean with less deeply nested photos? Are this photo's in root streams with only a few subfolders? Or are this photo's which are in the first subfolder of the root stream (and not a folder in a folder in a folder)?

Trying to get an idea how I should order my photo's for this app :)

**Comment by mvysny** [comment]
> But once all folders are completely loaded will the beginning of the random order be changed? Because the beginning of the display order was created when the random order was not completely random yet and because of that every time when the slideshow restarts the photo less deeply nested photos will be in the front of the random que.

Precisely. The algorithm is designed to show a picture as soon as possible. Since the dir crawling is breath-first, less deeply nested (closer to Root) photos are discovered sooner, so there is a high chance that the first photo will be one amongst such photos.

Let me rephrase:

> But once all folders are completely loaded will the beginning of the random order be changed? 

Yes and no. Newly discovered photos are added at random locations into the list (even to the beginning of the list). Yet, photos are slowly removed from the beginning of the list and displayed. Of course a photo can't "overtake" another photo that has already been shown.

If the algorithm was to be completely random, it would have to discover all photos first, and only then start with the slideshow. That would however introduce quite large initial delay, and thus I went the other way.

> What do you mean with less deeply nested photos? Are this photo's in root streams with only a few subfolders? Or are this photo's which are in the first subfolder of the root stream (and not a folder in a folder in a folder)?

Exactly. "Less deeply nested photos" is a moronic term I invented for the lack for better terms, for photos closer to the Root.

> Trying to get an idea how I should order my photo's for this app :)

Oh, I hope that you won't be forced to do that ;) If there is a delay long enough between the photos, then the dir crawling will finish quite soon. That means that first one or two photos are not picked completely randomly, but all consecutive photos should be picked randomly, and thus for a longer slideshows (with 15000 photos) and longer pauses between the photos this should work quite nicely.
